service/src/loader.rs

```diff
@@ -66,12 +74,35 @@ impl Loader {
         let mut service = self.service;
         let logger = Box::new(Logger::new(tx, deployment_id));
 
-        service.build(factory, logger).await?;
+        AssertUnwindSafe(service.build(factory, logger))
+            .catch_unwind()
+            .await
+            .unwrap_or(Err(Error::BuildPanic))?;
 
-        // Start service on this side of the FFI
-        let handle = tokio::spawn(async move { service.bind(addr)?.await? });
+        // channel used by task spawned below to indicate whether or not panic
+        // occurred in `service.bind` call
+        let (send, recv) = tokio::sync::oneshot::channel();
 
-        Ok((handle, self.so))
+        // Start service on this side of the FFI
+        let handle = tokio::spawn(async move {
+            let bound = AssertUnwindSafe(async { service.bind(addr) })
+                .catch_unwind()
+                .await;
+
+            send.send(bound.is_ok()).unwrap();
+
+            if let Ok(b) = bound {
+                b?.await?
+            } else {
+                Err(anyhow!("panic in `Service::bound`"))
+            }
+        });
+
+        if matches!(recv.await, Ok(true)) {
```

Comment:
Doesn't this block the `load()` function? Ie, the code in deployment.rs will `await` until this oneshot resolves?

Reply:
The implication being, does this block a second deploy?

Reply:
Ok, if `load()` returns immediately then I think this will be fine